### PR TITLE
Use explicit types in mailbox protocol

### PR DESCRIPTION
**Pull request by maxzhen**
#### Problem solved by the commit
After mailbox is supported on X3, the two ends of mailbox communication channel may run on different CPU architectures. It is important to use explicit type to avoid any mis-match in mailbox protocol data structures.
#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
N/A
#### How problem was solved, alternative solutions (if any) and why they were rejected
Use explicit types
#### Risks (if any) associated the changes in the commit
Should not change any data structure in binary
#### What has been tested and how, request additional testing if necessary
Tested on U55n with xbutil validate
#### Documentation impact (if any)
N/A

### CODE REVIEW

**Review by xuhz — APPROVED** [review]

**Review by arpitxilinx — APPROVED** [review]

**Review by richardtobin-xlnx — CHANGES_REQUESTED** [review]
There's still some silent padding:

https://godbolt.org/z/9aM67raoK

The above compiles the header for both 64-bit and 32-bit mode, and also compiles it with  structure packing enabled.  The sizeof() operator should give consistent results for all three compilations to ensure that all padding is present and accounted for.  It does not - the discrepancies are as follows:
 - `struct xcl_board_info` (1432 on both 64-bit and machines, but is 1430 when packed)
 - `struct xcl_sensor` (240 on 64-bit machine vs 236 on 32-bit machine)
 - `struct xcl_firewall` (104 on 64-bit machine vs 100 on 32-bit machine vs 98 when packed)
 - `struct xcl_sw_chan` (32 on 64-bit machine vs 28 on 32-bit machine)

Technically the 32-bit compilation is redundant for detecting the padding - I included it to show how it's inadequate.

Please fix up the listed structure definitions to make their tail padding explicit.

File: src/runtime_src/core/pcie/driver/linux/include/mailbox_proto.h [hunk 1]
```diff
@@ -338,22 +338,22 @@ struct xcl_mailbox_bitstream_kaddr {
...
+ * @req: opcode (enum xcl_mailbox_request)
  * @flags: flags of this message
  * @data: payload of variable length
  */
 struct xcl_mailbox_req {
 	uint64_t flags;
-	enum xcl_mailbox_request req;
-	char data[1]; /* variable length of payload */
+	int32_t req;
+	int8_t data[4]; /* variable length of payload, explicitly padded to int32_t */
```
Comment (richardtobin-xlnx) [inline src/runtime_src/core/pcie/driver/linux/include/mailbox_proto.h:356]: If I was doing this, I would have just added the padding as a separate definition - same comment for `struct xcl_sw_chan`.

Comment (maxzhen) [reply src/runtime_src/core/pcie/driver/linux/include/mailbox_proto.h]: Hi @richardtobin-xlnx ,

This is what I initially thought about. But if I keep "data" as the last member, it is more consistent with the fact that it is the starting point of a variable length chunk of payload in the same data structure buffer. Adding another member after it confuses me. Anyway, I changed the type (vs making it an array) but still keep it as the last one.

I also fixed the other data structures that you pointed out. Thanks for pointing me to the compiler tool. It was very helpful. Now I can see that all data structure size are the same without or without packing on both 32 and 64 bit platforms.

Please take another look.

**Comment by maxzhen** [comment]
retest this please - known issue.

**Comment by maxzhen** [comment]
retest this please

**Review by richardtobin-xlnx — APPROVED** [review]
Looks ok now, thanks

**Pull request merged**